presto-main/src/main/java/com/facebook/presto/execution/scheduler/SectionExecutionFactory.java

```diff
@@ -282,10 +302,52 @@ private StageScheduler createStageScheduler(
             if (isInternalSystemConnector(connectorId)) {
                 connectorId = null;
             }
-
-            NodeSelector nodeSelector = nodeScheduler.createNodeSelector(session, connectorId, maxTasksPerStage);
+            //filter nodes based on pool type, TODO make it configurable
+            NodeSelector nodeSelector = nodeScheduler.createNodeSelector(session, connectorId, maxTasksPerStage, nodePredicate);
             SplitPlacementPolicy placementPolicy = new DynamicSplitPlacementPolicy(nodeSelector, stageExecution::getAllTasks);
 
+            if (plan.getFragment().isLeaf()) {
+                stageExecution.registerStageTaskRecoveryCallback(taskId -> {
+                    HttpRemoteTask remoteTask = stageExecution.getAllTasks().stream()
```

Comment:
This seems like a core part of the retry logic, so better variable names would help understanding. 

Also, just for my understanding, is stageExecution.getAllTasks only the tasks for this leaf stage (ie for the plan.getFragment) ? Or is it all of the tasks across the entire query ?